cpp/src/arrow/compute/kernels/scalar_temporal_benchmark.cc

```diff
@@ -0,0 +1,149 @@
+auto non_zoned = timestamp(TimeUnit::NANO);
+
+#define DECLARE_TEMPORAL_ROUNDING_BENCHMARKS(OPTIONS)                              \
+  BENCHMARK_TEMPLATE(BenchmarkTemporalRounding, CeilTemporal, zoned, OPTIONS)      \
+      ->Apply(SetArgs);                                                            \
+  BENCHMARK_TEMPLATE(BenchmarkTemporalRounding, FloorTemporal, zoned, OPTIONS)     \
+      ->Apply(SetArgs);                                                            \
+  BENCHMARK_TEMPLATE(BenchmarkTemporalRounding, RoundTemporal, zoned, OPTIONS)     \
+      ->Apply(SetArgs);                                                            \
+  BENCHMARK_TEMPLATE(BenchmarkTemporalRounding, CeilTemporal, non_zoned, OPTIONS)  \
+      ->Apply(SetArgs);                                                            \
+  BENCHMARK_TEMPLATE(BenchmarkTemporalRounding, FloorTemporal, non_zoned, OPTIONS) \
+      ->Apply(SetArgs);                                                            \
+  BENCHMARK_TEMPLATE(BenchmarkTemporalRounding, RoundTemporal, non_zoned, OPTIONS) \
+      ->Apply(SetArgs);
+
+#define DECLARE_TEMPORAL_BENCHMARKS(OP)                                 \
+  BENCHMARK_TEMPLATE(BenchmarkTemporal, OP, non_zoned)->Apply(SetArgs); \
+  BENCHMARK_TEMPLATE(BenchmarkTemporal, OP, zoned)->Apply(SetArgs);
+
+#define DECLARE_TEMPORAL_BENCHMARKS_ZONED(OP) \
+  BENCHMARK_TEMPLATE(BenchmarkTemporal, OP, zoned)->Apply(SetArgs);
+
+// Temporal rounding benchmarks
+auto round_1_minute = RoundTemporalOptions(1, CalendarUnit::MINUTE);
+auto round_10_minute = RoundTemporalOptions(10, CalendarUnit::MINUTE);
+auto round_1_week = RoundTemporalOptions(1, CalendarUnit::WEEK);
+auto round_10_week = RoundTemporalOptions(10, CalendarUnit::WEEK);
+auto round_1_month = RoundTemporalOptions(1, CalendarUnit::MONTH);
+auto round_10_month = RoundTemporalOptions(10, CalendarUnit::MONTH);
+
+DECLARE_TEMPORAL_ROUNDING_BENCHMARKS(round_1_minute);
+DECLARE_TEMPORAL_ROUNDING_BENCHMARKS(round_1_week);
+DECLARE_TEMPORAL_ROUNDING_BENCHMARKS(round_1_month);
+DECLARE_TEMPORAL_ROUNDING_BENCHMARKS(round_10_minute);
+DECLARE_TEMPORAL_ROUNDING_BENCHMARKS(round_10_week);
+DECLARE_TEMPORAL_ROUNDING_BENCHMARKS(round_10_month);
+
+// Temporal component extraction
+DECLARE_TEMPORAL_BENCHMARKS(Year);
+DECLARE_TEMPORAL_BENCHMARKS(IsLeapYear);
+DECLARE_TEMPORAL_BENCHMARKS(Month);
+DECLARE_TEMPORAL_BENCHMARKS(Day);
+DECLARE_TEMPORAL_BENCHMARKS(DayOfYear);
+DECLARE_TEMPORAL_BENCHMARKS_ZONED(IsDaylightSavings);
+DECLARE_TEMPORAL_BENCHMARKS(USYear);
+DECLARE_TEMPORAL_BENCHMARKS(ISOYear);
+DECLARE_TEMPORAL_BENCHMARKS(ISOWeek);
+DECLARE_TEMPORAL_BENCHMARKS(USWeek);
+DECLARE_TEMPORAL_BENCHMARKS(Quarter);
+DECLARE_TEMPORAL_BENCHMARKS(Hour);
+DECLARE_TEMPORAL_BENCHMARKS(Minute);
+DECLARE_TEMPORAL_BENCHMARKS(Second);
+DECLARE_TEMPORAL_BENCHMARKS(Millisecond);
+DECLARE_TEMPORAL_BENCHMARKS(Microsecond);
+DECLARE_TEMPORAL_BENCHMARKS(Nanosecond);
+DECLARE_TEMPORAL_BENCHMARKS(Subsecond);
```

Comment:
@lidavidm do you think this is an excessive amount of benchmarks or should we have more?

Reply:
I think this is reasonable.